LiveSPICEVst/SimulationProcessor.cs

```diff
@@ -0,0 +1,322 @@
+
+                        buttonGroups[button.Group] = wrapper;
+
+                        InteractiveComponents.Add(wrapper);
+                    }
+
+                    wrapper.AddButton(button);
+                }
+            }
+
+            needRebuild = true;
+        }
+
+        /// <summary>
+        /// Run the cicuite simulation on a buffer of audio samples
+        /// </summary>
+        /// <param name="audioInput">Array of input samples</param>
+        /// <param name="audioOutput">Array of output samples</param>
+        public void RunSimulation(double[] audioInput, double[] audioOutput)
+        {
+            if (simulation == null)
+            {
+                if (needRebuild)
+                {
+                    UpdateSimulation(needRebuild);
+
+                    needRebuild = false;
+                }
+            }
+
+            if ((circuit == null) || (simulation == null))
+            {
+                audioInput.CopyTo(audioOutput, 0);
+            }
+            else
+            {
+                bool needUpdate = false;
+
+                lock (sync)
+                {
+                    foreach (ComponentWrapper component in InteractiveComponents)
+                    {
+                        if (component.NeedUpdate)
+                        {
+                            needUpdate = true;
+
+                            component.NeedUpdate = false;
+                        }
+
+                        if (component.NeedRebuild)
+                        {
+                            needRebuild = true;
+
+                            component.NeedRebuild = false;
+                        }
+                    }
+
+                    if (needUpdate || needRebuild)
+                    {
+                        UpdateSimulation(needRebuild);
+
+                        needRebuild = false;
+                    }
+
+                    simulation.Run(audioInput, audioOutput);
+                }
+            }
+        }
+
+        int clock = -1;
+        int update = 0;
+        TaskScheduler scheduler = new RedundantTaskScheduler(1);
+        object sync = new object();
+
+        /// <summary>
+        /// Update the simulation asynchronously
+        /// </summary>
+        /// <param name="rebuild">Whether a full simulation rebuild is required</param>
+        void UpdateSimulation(bool rebuild)
+        {
+            int id = Interlocked.Increment(ref update);
+            new Task(() =>
+            {
+                Analysis analysis = circuit.Analyze();
+                TransientSolution ts = TransientSolution.Solve(analysis, (Real)1 / (sampleRate * oversample));
+
+                lock (sync)
+                {
+                    if (id > clock)
+                    {
+                        if (rebuild)
+                        {
+                            Expression outputExpression = 0;
+
+                            foreach (Circuit.Component i in circuit.Components)
```

Comment:
I just tried this in the tests (it does something similar). OfType works, Sum does not.

Reply:
I've switched it to:

`Output = circuit.Components.OfType<Speaker>().Select(i => i.V) // voltage drop across the speakers`

Reply:
This isn't quite the same thing, because it will make an output for each speaker, instead of summing all the speakers. Maybe it is better to just generate an error in the case of multiple speakers though?

Reply:
I changed this back to doing a summation over the speakers. I throw an error if there is no input or no output. I did some (slightly ugly) work to be able to grab and display exceptions generated by the simulation update/rebuild in the background task.